src/ModelContextProtocol/Server/DestinationBoundMcpServer.cs

```diff
@@ -0,0 +1,34 @@
+﻿using ModelContextProtocol.Protocol.Messages;
+using ModelContextProtocol.Protocol.Transport;
+using ModelContextProtocol.Protocol.Types;
+
+namespace ModelContextProtocol.Server;
+
+internal sealed class DestinationBoundMcpServer(McpServer server, ITransport? transport) : IMcpServer
+{
+    public string EndpointName => server.EndpointName;
+    public ClientCapabilities? ClientCapabilities => server.ClientCapabilities;
+    public Implementation? ClientInfo => server.ClientInfo;
+    public McpServerOptions ServerOptions => server.ServerOptions;
+    public IServiceProvider? Services => server.Services;
+    public LoggingLevel? LoggingLevel => server.LoggingLevel;
+
+    public ValueTask DisposeAsync() => server.DisposeAsync();
+
+    public IAsyncDisposable RegisterNotificationHandler(string method, Func<JsonRpcNotification, CancellationToken, ValueTask> handler) => server.RegisterNotificationHandler(method, handler);
+
+    // This will throw because the server must already be running for this class to be constructed, but it should give us a good Exception message.
+    public Task RunAsync(CancellationToken cancellationToken) => server.RunAsync(cancellationToken);
+
+    public Task SendMessageAsync(JsonRpcMessage message, CancellationToken cancellationToken = default)
+    {
+        message.RelatedTransport = transport;
```

Comment:
Presumably it's ok that we don't unset this after the delegated call completes?

Should we assert it's already null before setting it?

Reply:
If we wanted to be really strict with what RelatedTransport was used for, I'd split it up into two properties: SourceTransport (which would only be on JsonRpcRequests) and DestinationTransport (which would be on all JsonRpcMessage types).

No single JsonRpcMessage needs to have both a SourceTransport and DestinationTransport configured, so it was easy to combine them. This is similar to what the typescript-sdk does with it relatedRequestId. I almost wanted to make this internal, so I really didn't want to add two properties if I could avoid it.

> Should we assert it's already null before setting it?

Sure. I'll add a Debug.Assert.